src/NCalc/Expression.cs

```diff
@@ -121,11 +121,28 @@ public static LogicalExpression Compile(string expression, bool nocache)
                 var lexer = new NCalcLexer(CharStreams.fromString(expression));
                 var lexerErrorListener = new ErrorListener<int>();
                 lexer.AddErrorListener(lexerErrorListener);
+
                 var parser = new NCalcParser(new CommonTokenStream(lexer));
+                parser.Interpreter.PredictionMode = PredictionMode.SLL;
                 var parserErrorListener = new ErrorListener<IToken>();
-                parser.AddErrorListener(parserErrorListener);
 
-                logicalExpression = parser.ncalcExpression().value;
+                parser.RemoveErrorListeners();
+                parser.ErrorHandler = new BailErrorStrategy();
+
+                try
+                {
+                    logicalExpression = parser.ncalcExpression().value;
+                }
+                catch(ParseCanceledException)
+                {
+                    lexer.Reset();
+
+                    parser.Reset();

```

Comment:
I am curious if/when this fallback mechanism is hit. Have you seen this being used for any of the test cases?

Reply:
The code is hit any time there is a parser error.  The test ShouldDetectSyntaxErrorsBeforeEvaluation hits the catch, but it doesn't test a successful fallback to the LL prediction mode.   I tried several complex expressions, but I wasn't able to find anything that failed on SLL and then succeeded on LL.  I'm not convinced the NCalc grammar is complex enough to need anything more than SLL, but I don't know how to prove it.  

The SLL parser is generally faster, and the LL fallback seems prudent.